difftest/dpi_t1rocketemu/src/interconnect.rs

```diff
@@ -62,68 +112,424 @@ pub trait RegDevice {
+      self.sys.tick();
+      self.dram_tick += 1;
+    }
+  }
+}
+
+impl MemoryModel for Arc<Mutex<DRAMModel>> {
+  fn push(&mut self, req: MemIdent) {
+    self.lock().unwrap().push(req);
+  }
+
+  fn pop(&mut self) -> Option<MemIdent> {
+    self.lock().unwrap().pop()
+  }
+
+  fn tick(&mut self) {
+    self.lock().unwrap().tick()
+  }
+}
+
+/// A trivial memory model, where all requests are immediately resolved
+/// and served in the FIFO order
+#[derive(Default)]
+pub struct TrivialModel {
+  holding: VecDeque<MemIdent>,
+}
+
+impl MemoryModel for TrivialModel {
+  fn push(&mut self, req: MemIdent) {
+    self.holding.push_back(req);
+  }
+  fn pop(&mut self) -> Option<MemIdent> {
+    self.holding.pop_front()
+  }
+  fn tick(&mut self) {}
+}
+
+/// Repersents a bulk memory device, with its memory model
+pub struct RegularMemory<M: MemoryModel> {
   data: Vec<u8>,
+  model: M,
 }
 
-impl RegularMemory {
-  pub fn with_size(size: u32) -> Self {
-    RegularMemory { data: vec![0; size as usize] }
+impl RegularMemory<TrivialModel> {
+  pub fn with_content(data: Vec<u8>) -> Self {
+    RegularMemory { data, model: TrivialModel::default() }
   }
 }
 
-impl Device for RegularMemory {
-  fn mem_read(&mut self, addr: AddrInfo, data: &mut [u8]) {
-    let mem_data = &self.data[addr.as_range()];
-    data.copy_from_slice(mem_data);
+impl RegularMemory<Arc<Mutex<DRAMModel>>> {
+  pub fn with_content_and_model(
+    data: Vec<u8>,
+    ds_cfg: impl AsRef<Path>,
+    ds_path: impl AsRef<Path>,
+  ) -> Self {
+    RegularMemory {
+      data,
+      model: Arc::new(Mutex::new(DRAMModel::new(ds_cfg, ds_path))),
+    }
   }
+}
 
-  fn mem_write(&mut self, addr: AddrInfo, data: &[u8]) {
-    let mem_data = &mut self.data[addr.as_range()];
-    mem_data.copy_from_slice(data);
+impl<M: MemoryModel + Send + Sync + 'static> RegularMemory<M> {
+  fn execute_read(&mut self, addr: AddrInfo) -> &[u8] {
+    &self.data[addr.as_range()]
   }
 
-  fn mem_write_masked(&mut self, addr: AddrInfo, data: &[u8], mask: &[bool]) {
+  fn execute_write(&mut self, addr: AddrInfo, data: &[u8], mask: Option<&[bool]>) {
     let mem_data = &mut self.data[addr.as_range()];
     memcpy_mask(mem_data, data, mask);
   }
 }
 
-fn memcpy_mask(dst: &mut [u8], src: &[u8], mask: &[bool]) {
-  for i in 0..mask.len() {
-    if mask[i] {
+impl<M: MemoryModel + Send + Sync + 'static> Device for RegularMemory<M> {
+  fn req(&mut self, req: MemReq<'_>) -> bool {
+    // dbg!(&req);
+    let ident = MemIdent {
+      id: req.id,
+      req: req.addr,
+      is_write: req.payload.is_write(),
+    };
+    self.model.push(ident);
+
+    if let MemReqPayload::Write(data, mask) = req.payload {
+      self.execute_write(req.addr, data, mask);
+    }
+    true
+  }
+
+  fn resp(&mut self) -> Option<MemResp<'_>> {
+    let popped = self.model.pop()?;
+    // dbg!(&popped);
+
+    // Construct MemResp
+    let payload = if popped.is_write {
+      MemRespPayload::WriteAck
+    } else {
```

Comment:
It seems to may reorder requests. W->W & W->R order is preserved. R->R reorder is harmless.

But it does allow reordering R->W to W->R. But we are modeling AXI, such order shall be handled by managers. Seems we are exploiting this behavior.

Is my understanding correct?

Reply:
I might not correctly understand the concern here. If what you are asking is that "where lies the responsibility for ordering concurrent overlapping R&W requests", then yes, it's the manager. More concretely, it's the iterating order of `incomplete_reads` and `incomplete_writes` in `src/drive.rs`. The device handles the requests as-is.

My understanding is that this reordering is allowed by AXI. The DPI side only need to guarantee any **consistent** order, s.t. the simulation result is deterministic. I'm even thinking of switch back to HashMap with a consistent hasher to get a little bit more coverage on potential memory request orderings.

Reply:
OK, I just want to confirm this. I have no concerns. I agree ensuring determinism is enough. I'm just a bit puzzled since no comments mention it. Comments like "It follows AXI ordering rules. No ordering between reads & writes are guaranteed by devices " with `MemReq` helps to clarify it.

In the future, I consider adding functions to detect R/W address overlaps from the same master, since it's very likely a bug of manager. This is not a blocker for this PR.